README.md
A library for adding many convenience methods / getters as extensions to commonly used dart classes. 

# Iterable < int >
### BigInt get sum
* Calculates the sum of all the elements in the `Iterable`. 
* **Runtime**: *O(n)* where *n* is the size of the `Iterable`

### BigInt get product
* Calculates the product of all the elements in the `Iterable`.
* **Runtime**: *O(n\*m)* where *n* is the number of elements in the `Iterable`, 
and *m* is the complexity of `BigInt`'s `*` operation (unknown at the moment)

### IntSummaryStatistics get stats
* Returns the summary stats for this iterable. Stats include,
sum, product, max, min, count, average, and range
* **Runtime**: *O(n\*m)* where *n* is the number of elements in the `Iterable`, 
  and *m* is the complexity of `BigInt`'s `*` operation (unknown at the moment)
* This is computed in one pass through the `Iterable`'s elements

## Usage
A simple usage example:

```dart
import 'package:better/better.dart';

main() {
  // Stats for list of ints
  print('\n\nStats for list of ints\n=======================================================================================');
  var r = Random();
  var length = r.nextInt(1 << 12);
  print('generating list...');
  var list = [for(int i = 0; i < length; i++) r.nextInt(1 << 31)];
  print('list generated');
  print('calculating stats');
  var start = DateTime.now().millisecondsSinceEpoch;
  var stats = list.stats;
  var end = DateTime.now().millisecondsSinceEpoch;
  print('Finished calculating stats in ${end - start} ms');
  print(stats);

  // String generation & sorting with comparators
  print('\n\nString generation & sorting with comparators\n=======================================================================================');
  var numStrings = r.nextInt(20) + 1;
  var strings = [for(var i in numStrings.range) RandomStrings.alpha(Random().nextInt(10))];
  strings = RandomStrings.alphaList(count: 5, stringLength: 5, randomizeCount: true, randomizeLength: true);
  print('Unsorted: $strings');
  print('Sorted: ${strings..sort(Strings.shortestFirst)}');
  print('Sorted: ${strings..sort(Strings.shortestFirst.reversed)}');
  print('Sorted: ${strings..sort(Strings.longestFirst)}'); // same as above

  // Pattern Matching / Regex
  print('\n\nPattern Matching / Regex\n=======================================================================================');
  print(RegexUtils.isValidEmailAddress('logan.w.russell@gmail.com'));
  print('logan.w.russell@gmail.com'.isValidEmail);
  print('.@gmail.com'.isValidEmail);
  print(RegexUtils.isValidNAPhoneNumber('9375555555'));
  print('+19375555555'.isValidNorthAmeicanPhoneNumber);
  print('999999'.isValidNorthAmeicanPhoneNumber);
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/loganrussell48/better/issues


CHANGELOG.md
## 0.1.0-dev.5
- Added Regex for Email Addresses & NA Phone Numbers
- Added Random Generation methods (check example)

## 0.1.0-dev.4
- Updated readme
- Formatted code
- Conform to pub standards

## 0.1.0-dev.3
- export int_summary_stats

## 0.1.0-dev.2

- Updated package description
- Fixed semantic changes

## 0.1.0-dev.1

- Initial version


test/better_test.dart
import 'package:better/better.dart';
import 'package:test/test.dart';

void main() {
  group('BIterable<int>', () {
    late List<int> emptyList;
    late Set<int> emptySet;
    late BigInt res;
    setUp(() {
      emptyList = <int>[];
      emptySet = <int>{};
      res = emptyList.sum;
    });

    test('has sum 0 when empty', () {
      res = emptyList.sum;
      expect(res, equals(BigInt.zero));
      res = emptySet.sum;
      expect(res, equals(BigInt.zero));
    });

    test('has product 1 when empty', () {
      res = emptyList.product;
      expect(res, equals(BigInt.one));
      res = emptySet.product;
      expect(res, equals(BigInt.one));
    });
  });
}


lib/src/Utils/RegexUtils.dart
class RegexUtils {

  static RegExp phone_regex = RegExp(r'^(?:\+?1[-.●]?)?\(?([0-9]{3})\)?[-.●]?([0-9]{3})[-.●]?([0-9]{4})$');
  static RegExp email_regex = RegExp(_email_regex_string);

  static const _email_regex_string = r'''(?:[a-z0-9!#$%&'*+/=?^_`{|}~-]+(?:\.[a-z0-9!#$%&'*+/=?^_`{|}~-]+)*|"(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21\x23-\x5b\x5d-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])*")@(?:(?:[a-z0-9](?:[a-z0-9-]*[a-z0-9])?\.)+[a-z0-9](?:[a-z0-9-]*[a-z0-9])?|\[(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?|[a-z0-9-]*[a-z0-9]:(?:[\x01-\x08\x0b\x0c\x0e-\x1f\x21-\x5a\x53-\x7f]|\\[\x01-\x09\x0b\x0c\x0e-\x7f])+)\])''';

  /// Returns true if [email] is a valid email address
  static bool isValidEmailAddress(String email) => email_regex.hasMatch(email);

  /// Returns true if [phone] is a valid North American Phone number
  /// The number must be contain the area code - i.e 7 digit numbers are not accepted
  static bool isValidNAPhoneNumber(String phone) => phone_regex.hasMatch(phone);

}

lib/src/int_summary_stats.dart
class IntSummaryStatistics {
  int _min = ~(1 << 63);
  int _max = 1 << 63;
  int _range = (1 << 63) - (~(1 << 63));
  int _count = 0;
  num _average = 0;
  BigInt _sum = BigInt.zero;
  BigInt _prod = BigInt.one;

  IntSummaryStatistics(Iterable<int> iterable) {
    if (iterable.isEmpty) return;
    var it = iterable.iterator;
    while (it.moveNext()) {
      _count++;
      _sum += BigInt.from(it.current);
      _prod *= BigInt.from(it.current);
      if (it.current < _min) _min = it.current;
      if (it.current > _max) _max = it.current;
    }
    _range = _max - _min;
    _average = _sum / BigInt.from(_count);
  }

  int get min => _min;

  int get max => _max;

  int get range => _range;

  int get count => _count;

  double get average => _average;

  BigInt get sum => _sum;

  BigInt get product => _prod;

  @override
  String toString() {
    return 'IntSummaryStatistics{min: $min, max: $max, range: $range, count: $count, average: $average, sum: $sum, prod: $product}';
  }
}


lib/src/num_summary_stats.dart
class NumSummaryStatistics {
  num _min = ~(1 << 63);
  num _max = 1 << 63;
  num _range = (1 << 63) - (~(1 << 63));
  int _count = 0;
  num _sum = 0;
  num _avg = 0;
  NumSummaryStatistics(Iterable<num> iterable) {
    if (iterable.isEmpty) return;
    var it = iterable.iterator;
    while (it.moveNext()) {
      _count++;
      if (it.current < _min) _min = it.current;
      if (it.current > _max) _max = it.current;
      _sum += it.current;
    }
    _range = _max - _min;
    _avg = _sum / count;
  }

  num get min => _min;

  num get max => _max;

  num get range => _range;

  int get count => _count;

  num get average => _avg;

  num get sum => _sum;

  @override
  String toString() {
    return 'IntSummaryStatistics{min: $min, max: $max, range: $range, count: $count, average: $average, sum: $sum}';
  }
}


lib/src/Extensions/better_base.dart
import 'dart:collection';
import 'dart:math';
import 'package:meta/meta.dart';
import 'package:better/src/Utils/RegexUtils.dart';
import 'package:better/src/num_summary_stats.dart';

import '../int_summary_stats.dart';

extension BIterableT<T> on Iterable<T> {
  int countWhere(bool Function(T) predicate) {
    var count = 0;
    var it = this.iterator;
    while (it.moveNext()) {
      if (predicate(it.current)) count++;
    }
    return count;
  }

  /// Performs the given [action] on each element, providing sequential index
  /// with the element.
  void forEachIndexed(void Function(T element, int index) action) {
    var index = 0;
    for (var element in this) {
      action(element, index++);
    }
  }
}
extension BIterableInt on Iterable<int> {
  static final String _rangeInputSizeErrorMessage =
      '''The iterable upon which the "range" getter was invoked has invalid size.
       Make sure there are only 2 num elements in the iterable.''';

  ///Returns the sum of the elements as a BigInt to avoid overflow
  BigInt get sum => fold(BigInt.zero, (l, r) => l + BigInt.from(r));

  ///Returns the product of the elements as a BigInt to avoid overflow
  BigInt get product => fold(BigInt.one, (l, r) => l * BigInt.from(r));

  ///Returns the summary stats for this iterable. Stats include,
  ///sum, product, max, min, count, and range
  IntSummaryStatistics get stats => IntSummaryStatistics(this);

  ///returns an iterable containing all the ints in the given range, inclusive
  ///
  /// Notice that the input is `num` not `int`.
  Iterable<int> get asRange {
    if (length != 2) throw ArgumentError(_rangeInputSizeErrorMessage);
    var it = iterator;
    var start = (it..moveNext()).current;
    var end = (it..moveNext()).current;
    return [for (int i = start.ceil(); i <= end; i++) i];
  }
}


extension BListInt on List<int> {
  List<int> get prefixSum {
    var clone = this.clone;
    clone.forEachIndexed((elem, ind) => clone[ind] += (ind >= 1) ? clone[ind-1] : 0);
    return clone;
  }

  List<int> get clone => [...this];

  List<int> get radixSorted {
    var q0 = Queue(), q1 = Queue();
    var bits = 32;
    for( var i = 0, mask = 1; i < bits; ++i, mask <<= 1 ) {
      var t0 = 0, t1 = 0;
      for (var i = 0; i < length; ++i ) {
        if ( (this[i] & mask) == 0 ) {
          q0.add(this[i]);
        } else {
          q1.add(this[i]);
        }
      }
      var i = 0;
      while(q0.isNotEmpty){
        this[i++] = q0.removeFirst();
      }
      while(q1.isNotEmpty){
        this[i++] = q1.removeFirst();
      }
    }
    return this;
  }
}

extension Ints on int {
  Iterable<int> get range => this < 0 ? [this, 0].asRange : [0, this].asRange;

  BigInt get factorial => [for(int i = 2; i <= this; i++) i].fold(BigInt.one, (l, r) => BigInt.from(r)*l);
}

extension BIterableNum on Iterable<num> {
  static final String _rangeInputSizeErrorMessage =
      '''The iterable upon which the "range" getter was invoked has invalid size.
       Make sure there are only 2 num elements in the iterable.''';

  ///Returns the sum of the elements as a BigInt to avoid overflow
  BigInt get sum => fold(BigInt.zero, (l, r) => l + BigInt.from(r));

  ///Returns the product of the elements as a BigInt to avoid overflow
  BigInt get product => fold(BigInt.one, (l, r) => l * BigInt.from(r));

  ///Returns the summary stats for this iterable. Stats include,
  ///sum, product, max, min, count, and range
  NumSummaryStatistics get stats => NumSummaryStatistics(this);

  ///returns an iterable containing all the ints in the given range, inclusive
  ///
  /// Notice that the input is `num` not `int`.
  Iterable<int> get range {
    if (length != 2) throw ArgumentError(_rangeInputSizeErrorMessage);
    var it = iterator;
    var start = (it..moveNext()).current;
    var end = (it..moveNext()).current;
    return [for (int i = start.ceil(); i <= end; i++) i];
  }
}

extension BIterableString on Iterable<String> {}



extension Strings on String {
  static Comparator<String> get shortestFirst => (a, b) => a.length - b.length;

  static Comparator<String> get longestFirst => (a, b) => b.length - a.length;
}

extension Chars on Random {
  static int get randAlpha => Random().nextInt(26) + 65 + Bit.random * 32;

  static int get randAlphaLower => Random().nextInt(26) + 97;

  static int get randAlphaUpper => Random().nextInt(26) + 65;

  static int get randDigit => Random().nextInt(10) + '0'.codeUnitAt(0);
}

extension RandomStrings on String {
  ///generates random chars that match [a-zA-Z]
  static String alpha(int length) {
    return String.fromCharCodes([for (var i in length.range) Chars.randAlpha]);
  }

  ///generates random chars that match [A-Z]
  static String alphaUpper(int length) {
    return String.fromCharCodes(
        [for (var i in length.range) Chars.randAlphaUpper]);
  }

  ///generates random chars that match [a-z]
  static String alphaLower(int length) {
    return String.fromCharCodes(
        [for (var i in length.range) Chars.randAlphaLower]);
  }

  ///generates random chars that match [0-9]
  static String numeric(int length) {
    return String.fromCharCodes([for (var i in length.range) Chars.randDigit]);
  }

  static List<String> alphaList(
      {@Required() int count,
      @Required() int stringLength,
      bool randomizeCount = false,
      bool randomizeLength = false}) {
    return [
      for (var i in (randomizeCount ? Random().nextInt(count) : count).range)
        RandomStrings.alpha(
            randomizeLength ? Random().nextInt(stringLength) : stringLength)
    ];
  }

  static List<String> alphaLowerList(
      {@Required() int count,
      @Required() int stringLength,
      bool randomizeCount = false,
      bool randomizeLength = false}) {
    return [
      for (var i in (randomizeCount ? Random().nextInt(count) : count).range)
        RandomStrings.alphaLower(
            randomizeLength ? Random().nextInt(stringLength) : stringLength)
    ];
  }

  static List<String> alphaUpperList(
      {@Required() int count,
      @Required() int stringLength,
      bool randomizeCount = false,
      bool randomizeLength = false}) {
    return [
      for (var i in (randomizeCount ? Random().nextInt(count) : count).range)
        RandomStrings.alphaUpper(
            randomizeLength ? Random().nextInt(stringLength) : stringLength)
    ];
  }

  static List<String> numericList(
      {@Required() int count,
      @Required() int stringLength,
      bool randomizeCount = false,
      bool randomizeLength = false}) {
    if (count < 0) {
      throw ArgumentError.value(count, 'count', 'Value should be >= 0');
    }
    if (stringLength < 0) {
      throw ArgumentError.value(
          stringLength, 'stringLength', 'value should be >= 0');
    }
    return [
      for (var i in (randomizeCount ? Random().nextInt(count) : count).range)
        RandomStrings.numeric(
            randomizeLength ? Random().nextInt(stringLength) : stringLength)
    ];
  }
}

extension Bool on bool {
  /// a lot of people don't know that Random has nextBool, so this will
  /// help with discovery
  static bool get flip => Random().nextBool();

  /// returns 1 if true, 0 if false
  int get asInt => this ? 1 : 0;
}

extension Bit on bool {
  static int get random => Random().nextBool().asInt;
}

extension Comparators<T> on Comparator<T> {
  /// Reverses a given comparator
  ///
  /// Note: Does NOT reverse the order of elements that are equal to the comparator
  Comparator<T> get reversed => (T a, T b) => -1 * this(a, b);
}

extension BIterableComparable<T extends Comparable> on Iterable<T> {
  T max(Comparator<T> cmp) {
    var it = iterator;
    var maximum = (it..moveNext()).current;
    var next;
    while (it.moveNext()) {
      next = it.current;
      maximum = (cmp(maximum, next) <= 0) ? maximum : next;
    }
    return maximum;
  }
}

extension StringUtils on String {
  bool get isValidEmail => RegexUtils.isValidEmailAddress(this);

  bool get isValidNorthAmericanPhoneNumber =>
      RegexUtils.isValidNAPhoneNumber(this);

  bool get isValidDateTime => DateTime.tryParse(this) != null;
}
//extension BListComparable<T> on List<T> {
//  bool isSorted({Comparator comparator}){
//    if(isEmpty){
//      return true;
//    }
//    if(comparator == null){
//      if(T is Comparable){
//        for(var i = 0; i < length - 1; i++){
//          if((this[i] ).compareTo(this[i+1]) > 0){
//            return false;
//          }
//        }
//      }
//    }
//  }
//}


example/better_example.dart
import 'dart:io';
import 'dart:math' as Math;

import 'package:better/better.dart';
import 'package:better/src/Utils/RegexUtils.dart';
import 'package:intl/intl.dart';

//TODO: Changes here should also be made to the readme
void main() {
  // Stats for list of ints
  print('\n\nStats for list of ints\n=======================================================================================');
  // Generates 20 random numeric strings with length 0-3 digits long
  // Then mapped to ints
  var gen = RandomStrings.numericList(count: 20, stringLength: 3, randomizeLength: true).map(int.parse).toList();
  print(gen.stats);
  var r = Math.Random();
  var length = r.nextInt(1 << 12);
  print('generating list...');
  var list = [for(int i = 0; i < length; i++) r.nextInt(1 << 31)];
  print('list generated');
  print('calculating stats');
  var start = DateTime.now().millisecondsSinceEpoch;
  var stats = list.stats;
  var end = DateTime.now().millisecondsSinceEpoch;
  print('Finished calculating stats in ${end - start} ms');
  print('keep in mind that the product of all the values may be huge - thousands of digits long');
  print(stats);
  print('number of digits in product: ${NumberFormat(',###').format(stats.product.toString().length)}');

  // String generation & sorting with comparators
  print('\n\nString generation & sorting with comparators\n=======================================================================================');
  var numStrings = r.nextInt(20) + 1;
  var strings = [for(var i in numStrings.range) RandomStrings.alpha(Math.Random().nextInt(10))];
  strings = RandomStrings.alphaList(count: 7, stringLength: 8, randomizeCount: false, randomizeLength: true);
  print('Unsorted: \t$strings (scrambled list, as they were generated)');
  print('Sorted: \t${strings..sort(Strings.shortestFirst)} (shortest first)');
  print('Sorted: \t${strings..sort(Strings.shortestFirst.reversed)} (shortest first, then reversed)');
  print('Sorted: \t${strings..sort(Strings.longestFirst)} (longest first, same as above)');

  // Pattern Matching / Regex
  print('\n\nPattern Matching / Regex\n=======================================================================================');

  print("RegexUtils.isValidEmailAddress('logan.w.russell@gmail.com') -> \t${RegexUtils.isValidEmailAddress('logan.w.russell@gmail.com')}");
  print("'logan.w.russell@gmail.com'.isValidEmail -> \t\t\t\t\t${'logan.w.russell@gmail.com'.isValidEmail}");
  print("'.@gmail.com'.isValidEmail -> \t\t\t\t\t\t\t\t\t${'.@gmail.com'.isValidEmail}");
  print("RegexUtils.isValidNAPhoneNumber('9375555555') -> \t\t\t\t${RegexUtils.isValidNAPhoneNumber('9375555555')}");
  print("'+19375555555'.isValidNorthAmericanPhoneNumber -> \t\t\t\t${'+19375555555'.isValidNorthAmericanPhoneNumber}");
  print("'999999'.isValidNorthAmericanPhoneNumber -> \t\t\t\t\t${'999999'.isValidNorthAmericanPhoneNumber}");

  print('\n\n\n');
  start = DateTime.now().microsecondsSinceEpoch;
  var val = 100;
  var fact = val.factorial;
  end = DateTime.now().microsecondsSinceEpoch;
  print('Finished calculating factorial in ${end - start} μs');
  print('$val! = $fact\n\n');

  var arr = [0,7,4,-6,2,1,8,6,4,9,2,5,4,6,8,5,4,3,0,0,0,0,0,5,1,2,3,6];
  print(arr.prefixSum);
  print(arr.reversed.toList().prefixSum);

  print('\n\n');
  gen = RandomStrings.numericList(count: 50000, stringLength: 10, randomizeLength: true).map(int.parse).toList();
  print('Unsorted: \t\t\t\t$gen');
  start = DateTime.now().microsecondsSinceEpoch;
  var radixSorted = gen.radixSorted;
  end = DateTime.now().microsecondsSinceEpoch;
  print('Binary Radix Sorted: \t$radixSorted in ${end - start} μs');
  start = DateTime.now().microsecondsSinceEpoch;
  var defaultSorted = gen..sort();
  end = DateTime.now().microsecondsSinceEpoch;
  print('Sorted:   \t\t\t\t$defaultSorted in ${end - start} μs');
}
